crates/extensions/c8y_http_proxy/src/actor.rs

```diff
@@ -162,80 +166,145 @@ impl C8YHttpProxyActor {
+                StatusCode::UNAUTHORIZED | StatusCode::FORBIDDEN => {
+                    self.try_request_with_fresh_token(build_request).await
+                }
+                StatusCode::NOT_FOUND => {
+                    self.try_request_with_fresh_internal_id(device_type, build_request)
+                        .await
+                }
+                code => Err(C8YRestError::FromHttpError(
+                    tedge_http_ext::HttpError::HttpStatusError(code),
+                )),
+            },
 
-        // TODO Add timeout
-        // TODO Manage 403 errors
-        let request = request_builder.build()?;
+            Err(e) => Err(C8YRestError::FromHttpError(e)),
+        }
+    }
+
+    async fn get_fresh_token(&mut self) -> Result<String, C8YRestError> {
+        self.end_point.token = None;
+        self.get_and_set_jwt_token().await
+    }
+
+    async fn try_request_with_fresh_token(
+        &mut self,
+        build_request: impl Fn(&C8yEndPoint) -> Result<HttpRequestBuilder, C8YRestError>,
+    ) -> Result<HttpResult, C8YRestError> {
+        // get new token not the cached one
+        self.get_fresh_token().await?;
+        // build the request
+        let request_builder = build_request(&self.end_point);
+        let request = request_builder?.build()?;
+        // retry the request
         Ok(self.peers.http.await_response(request).await?)
     }
 
-    async fn create_event(
+    async fn try_request_with_fresh_internal_id(
         &mut self,
-        mut c8y_event: C8yCreateEvent,
-    ) -> Result<EventId, C8YRestError> {
-        if c8y_event.source.is_none() {
-            c8y_event.source = Some(C8yManagedObject {
-                id: self.end_point.get_c8y_internal_id().to_string(),
-            });
-        }
-        self.send_event_internal(c8y_event).await
+        device_type: DeviceType,
+        build_request: impl Fn(&C8yEndPoint) -> Result<HttpRequestBuilder, C8YRestError>,
+    ) -> Result<HttpResult, C8YRestError> {
+        // get new internal id not the cached one
+        self.get_and_set_internal_id(device_type).await?;
+
+        let request_builder = build_request(&self.end_point);
+        let request = request_builder?
+            .bearer_auth(self.end_point.token.clone().unwrap_or_default())
+            .build()?;
+        Ok(self.peers.http.await_response(request).await?)
+    }
+
+    async fn create_event(&mut self, c8y_event: C8yCreateEvent) -> Result<EventId, C8YRestError> {
+        let create_event = |internal_id: String| -> C8yCreateEvent {
+            C8yCreateEvent {
+                source: Some(C8yManagedObject { id: internal_id }),
+                event_type: c8y_event.event_type.clone(),
+                time: c8y_event.time,
+                text: c8y_event.text.clone(),
+                extras: c8y_event.extras.clone(),
+            }
+        };
+        let device_type = self.end_point.get_device_type(None);
```

Comment:
The `device_type` will always be the main device? As you're passing `None` here. Why? What if an event has to be created for a child device?

Reply:
Yes, this function `create_event` is only used by the `main` device. If a child device wants to use this function then it has to pass the `device` id, then the `DeviceType` can be derived out of this.

Reply:
That doesn't sound right. Even if a child device sends an event with extra fragments, it has to go through this same API call. So, we're missing something here.